unit test flake: race in k8s.io/apiserver/pkg/storage/cache

```
Executing tests from //staging/src/k8s.io/apiserver/pkg/storage/cacher:go_default_test
-----------------------------------------------------------------------------
==================
WARNING: DATA RACE
Read at 0x00c00016a238 by goroutine 35:
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.(*Cacher).dispatchEvent()
      staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:792 +0xb1
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.TestDispatchingBookmarkEventsWithConcurrentStop.func1()
      staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:681 +0x42

Previous write at 0x00c00016a238 by goroutine 39:
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.(*Cacher).startDispatching()
      staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:833 +0x121
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.(*Cacher).dispatchEvent()
      staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:786 +0x46
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.TestDispatchingBookmarkEventsWithConcurrentStop.func1()
      staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:681 +0x42

Goroutine 35 (running) created at:
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.TestDispatchingBookmarkEventsWithConcurrentStop()
      staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:680 +0x88b
  testing.tRunner()
      GOROOT/src/testing/testing.go:865 +0x163

Goroutine 39 (finished) created at:
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.TestDispatchingBookmarkEventsWithConcurrentStop()
      staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:680 +0x88b
  testing.tRunner()
      GOROOT/src/testing/testing.go:865 +0x163
==================
--- FAIL: TestDispatchingBookmarkEventsWithConcurrentStop (0.61s)
    testing.go:809: race detected during execution of test
```

cc @wojtek-t 

ref #73958

--- a/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go
@@ -25,7 +25,7 @@ import (
 	"testing"
 	"time"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/meta"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/fields"
@@ -678,12 +678,16 @@ func TestDispatchingBookmarkEventsWithConcurrentStop(t *testing.T) {
 			t.Fatalf("failure to update version of object (%d) %#v", bookmark.ResourceVersion, bookmark.Object)
 		}
 
+		wg := sync.WaitGroup{}
+		wg.Add(2)
 		go func() {
 			cacher.dispatchEvent(bookmark)
+			wg.Done()
 		}()
 
 		go func() {
 			w.Stop()
+			wg.Done()
 		}()
 
 		done := make(chan struct{})
@@ -700,5 +704,6 @@ func TestDispatchingBookmarkEventsWithConcurrentStop(t *testing.T) {
 			t.Fatal("receive result timeout")
 		}
 		w.Stop()
+		wg.Wait()
 	}
 }

